desc/compute/utils.py

```diff
@@ -1347,3 +1360,776 @@ def body(i, mins):
+
+    Returns
+    -------
+    intersect, is_bp, bp1, bp2 : ndarray, ndarray, ndarray, ndarray
+        The polynomials' intersection points with 1 / λ is given by ``intersect``.
+        In order to be JIT compilable, the returned array must have a shape that
+        accommodates the case where each cubic polynomial intersects 1 / λ thrice.
+        The returned array ``intersect`` has shape (pitch.size, M * L, N, NUM_ROOTS).
+
+        The boolean mask ``is_bp`` encodes whether a given entry in ``intersect``
+        is a valid bounce point. The boolean masks ``bp1`` and ``bp2`` encode whether
+        a given entry in ``intersect`` is a valid starting and ending bounce point,
+        respectively. These arrays have shape (pitch.size * M * L, N * NUM_ROOTS).
+
+    """
+    ML = poly_B.shape[1]
+    N = poly_B.shape[2]
+    NUM_ROOTS = 3
+    a_min = zeta[:-1]
+    a_max = zeta[1:]
+    intersect = cubic_poly_roots(poly_B, 1 / pitch, a_min, a_max, sort=True).reshape(
+        pitch.size, ML, N, NUM_ROOTS
+    )
+    # Typically, this is the shape that is needed later, (as is the case in
+    # bounce_integrals because jax.vmap works over one axis at a time), so we
+    # reshape early here to prevent unnecessary data movement.
+    B_z = polyval(intersect, poly_B_z[:, jnp.newaxis]).reshape(
+        pitch.size * ML, N * NUM_ROOTS
+    )
+    # Check sign of derivative to determine whether root is a valid bounce point.
+    bp1 = B_z <= 0
+    bp2 = B_z >= 0
+    # Periodic boundary to compute bounce integrals of particles trapped outside
+    # this snapshot of the field lines.
+    is_bp = bp1 & jnp.roll(bp2, -1, axis=-1)
+    bp1 = bp1 & is_bp
+    bp2 = bp2 & is_bp
+    return intersect, is_bp, bp1, bp2
+
+
+def _compute_bp_if_given_pitch(pitch, zeta, poly_B, poly_B_z, *original, err=False):
+    """Return the quantities needed by the ``bounce_integrals`` function.
+
+    Parameters
+    ----------
+    pitch : ndarray
+        λ values representing the constant function 1 / λ.
+        If None, returns the given ``original`` tuple.
+    zeta : ndarray
+        Field line-following ζ coordinates of spline knots.
+    poly_B : ndarray
+        Polynomial coefficients of the cubic spline of |B|.
+    poly_B_z : ndarray
+        Polynomial coefficients of the cubic spline of ∂|B|/∂_ζ.
+    original : tuple
+        pitch, intersect, is_bp, bp1, bp2.
+    err : bool
+        Whether to raise an error if ``pitch`` is None and ``original`` is empty.
+
+    Returns
+    -------
+    output : tuple
+        (pitch, intersect, is_bp, bp1, bp2).
+
+    """
+    if pitch is None:
+        if err and not original:
+            raise ValueError("No pitch values were given.")
+        return original
+    else:
+        pitch = jnp.atleast_1d(pitch)
+        intersect, is_bp, bp1, bp2 = _get_bounce_points(pitch, zeta, poly_B, poly_B_z)
+        return pitch, intersect, is_bp, bp1, bp2
+
+
+def bounce_integral(
+    eq,
```

Comment:
we should think about if/how this will fit with the existing `compute` API which doesn't know anything about the `Equilibrium` class.